src/coprocessor/dag/dag.rs

```diff
@@ -26,52 +25,56 @@ use coprocessor::{Error, Result};
 use coprocessor::endpoint::{get_chunk, get_pk, to_pb_error, ReqContext};
 use storage::{Snapshot, SnapshotStore, Statistics};
 
-use super::executor::{AggregationExecutor, Executor as DAGExecutor, IndexScanExecutor,
-                      LimitExecutor, Row, SelectionExecutor, TableScanExecutor, TopNExecutor};
+use super::executor::{build_exec, Executor, Row};
 
-pub struct DAGContext<'s> {
-    columns: Rc<Vec<ColumnInfo>>,
+pub struct DAGContext {
+    columns: Arc<Vec<ColumnInfo>>,
     has_aggr: bool,
-    req: DAGRequest,
-    ranges: Vec<KeyRange>,
-    snap: &'s Snapshot,
-    eval_ctx: Rc<EvalContext>,
-    req_ctx: &'s ReqContext,
+    req_ctx: Arc<ReqContext>,
+    exec: Box<Executor>,
+    output_offsets: Vec<u32>,
 }
 
-impl<'s> DAGContext<'s> {
+impl DAGContext {
     pub fn new(
-        req: DAGRequest,
+        mut req: DAGRequest,
         ranges: Vec<KeyRange>,
-        snap: &'s Snapshot,
-        eval_ctx: Rc<EvalContext>,
-        req_ctx: &'s ReqContext,
-    ) -> DAGContext<'s> {
-        DAGContext {
-            req: req,
-            columns: Rc::new(vec![]),
-            ranges: ranges,
-            snap: snap,
-            has_aggr: false,
-            eval_ctx: eval_ctx,
+        snap: Box<Snapshot>,
```

Comment:
can we avoid Box here?

Reply:
It's hard, because `TiDbEndPoint` uses `DAGContext`, and there is also `Box<Snapshot>` in `TiDbEndPoint`, too. So we need to update many places, maybe do it in a seperate PR is better.